backend/msgpack/queue_mgr.go

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/pool"
+)
+
+var (
+	errInstanceQueueManagerClosed = errors.New("instance queue manager closed")
+)
+
+// instanceQueueManager manages instance queues.
+type instanceQueueManger interface {
+	// AddInstances adds instances.
+	AddInstances(instances []instance) error
+
+	// RemoveInstances removes instances.
+	RemoveInstances(instances []instance) error
+
+	// WriteTo writes a msgpack encoded buffer to the target instances.
+	WriteTo(instances []instance, buf msgpack.Buffer) error
+
+	// Close closes the connection manager.
+	Close() error
+}
+
+type queueManager struct {
+	sync.RWMutex
+
+	opts        ServerOptions
+	writeOpPool *writeOpPool
+	queues      map[string]*refCountedQueue
+	closed      bool
+}
+
+func newInstanceQueueManager(opts ServerOptions) instanceQueueManger {
+	writeOpPoolOpts := pool.NewObjectPoolOptions().SetSize(opts.WriteOpPoolSize())
+	writeOpPool := newWriteOpPool(writeOpPoolOpts)
+	writeOpPool.Init()
+
+	return &queueManager{
+		opts:        opts,
+		writeOpPool: writeOpPool,
+		queues:      make(map[string]*refCountedQueue),
+	}
+}
+
+func (mgr *queueManager) AddInstances(instances []instance) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.closed {
+		return errInstanceQueueManagerClosed
+	}
+	for _, instance := range instances {
+		id := instance.ID()
+		queue, exists := mgr.queues[id]
+		if !exists {
+			queue = newRefCountedQueue(instance, mgr.opts)
+			mgr.queues[id] = queue
+		}
+		queue.IncRef()
```

Comment:
because the newActivePlacement added the instances for current and future placement snapshots, the ref count helps with expiring old snapshots when new snapshot became active.